File: packages/bridge/src/source.ts
import { invoke } from './invoke.js';

/**
 * - builtin: Built-in bundle which is included in the application.
 * - remote: Remote bundle which is downloaded from a remote server.
 */
export type SourceKind = 'builtin' | 'remote';

/** What the manifest records for one version of a bundle. */
export interface ManifestVersionData {
  /** Integrity value to verify the bundle. */
  integrity?: string;
  /** Arbitrary string-valued metadata the update carried for this version. */
  metadata?: Record<string, string>;
}

/** Where a version stands in its bundle's lifecycle. */
export type ManifestBundleItemStatus = 'current' | 'previous' | 'staged' | 'orphan';

/** One version of one bundle, as the manifest records it. */
export interface ManifestBundleItem {
  /** The name of the bundle. */
  name: string;
  /** The version of the bundle. */
  version: string;
  /** Whether this version is the current/previous/staged one, or an orphan. */
  status: ManifestBundleItemStatus;
  /** Bundle manifest data. */
  data: ManifestVersionData;
}

/** List item of bundles. */
export interface SourceListItem {
  /** Which source (builtin or remote) records this item. */
  source: SourceKind;
  /** The bundle item. */
  item: ManifestBundleItem;
}

/** Bundle version with the source that provides it. */
export interface BundleSourceVersion {
  /** The source that provides this version. */
  source: SourceKind;
  /** The version of the bundle. */
  version: string;
}

export type ManifestSetCurrentVersionResultKind =
  /** The version is now the current one. */
  | 'settled'
  /** The bundle does not exist in the manifest. */
  | 'not_exists'
  /** The version does not exist in the manifest. */
  | 'version_not_exists';

export interface ManifestSetCurrentVersionResult {
  name: string;
  version: string;
  kind: ManifestSetCurrentVersionResultKind;
}

/** The version to stage, and what the manifest should record for it. */
export interface ManifestStageData {
  version: string;
  data?: ManifestVersionData;
}

export type ManifestStageResultKind =
  /** The version is now the staged one. */
  | 'staged'
  /** The version is the current one, so it cannot be staged. */
  | 'in_use';

export interface ManifestStageResult {
  name: string;
  version: string;
  kind: ManifestStageResultKind;
}

/** The versions to remove from one bundle. */
export interface ManifestRemoveData {
  versions: string[];
  /** Remove even the version in use. Default: `false`. */
  force?: boolean;
}

export type ManifestRemoveResultKind =
  | 'removed'
  /** The bundle does not exist in the manifest. */
  | 'not_exists'
  /** The version does not exist in the manifest. */
  | 'version_not_exists'
  /** The version is the current one; pass `force` to remove it anyway. */
  | 'in_use';

export interface ManifestRemoveResult {
  name: string;
  version: string;
  kind: ManifestRemoveResultKind;
}

export interface ManifestPruneResult {
  name: string;
  /** The orphan versions that were removed. */
  prunedVersions: string[];
}

async function listBundles(): Promise<SourceListItem[]> {
  return invoke<SourceListItem[]>('sourceListBundles');
}

async function listBuiltinBundles(): Promise<SourceListItem[]> {
  return invoke<SourceListItem[]>('sourceListBuiltinBundles');
}

async function listRemoteBundles(): Promise<SourceListItem[]> {
  return invoke<SourceListItem[]>('sourceListRemoteBundles');
}

async function getVersion(bundleName: string): Promise<BundleSourceVersion | null> {
  return invoke<BundleSourceVersion | null>('sourceGetVersion', { bundleName });
}

async function getRemoteStagedVersion(bundleName: string): Promise<string | null> {
  return invoke<string | null>('sourceGetRemoteStagedVersion', { bundleName });
}

async function getRemotePreviousVersion(bundleName: string): Promise<string | null> {
  return invoke<string | null>('sourceGetRemotePreviousVersion', { bundleName });
}

async function getBuiltinVersionData(
  bundleName: string,
  version: string
): Promise<ManifestVersionData | null> {
  return invoke<ManifestVersionData | null>('sourceGetBuiltinVersionData', {
    bundleName,
    version,
  });
}

async function getRemoteVersionData(
  bundleName: string,
  version: string
): Promise<ManifestVersionData | null> {
  return invoke<ManifestVersionData | null>('sourceGetRemoteVersionData', {
    bundleName,
    version,
  });
}

async function updateRemoteVersion(
  bundleName: string,
  version: string
): Promise<ManifestSetCurrentVersionResult> {
  return invoke<ManifestSetCurrentVersionResult>('sourceUpdateRemoteVersion', {
    bundleName,
    version,
  });
}

async function updateRemoteVersions(
  items: Record<string, string>
): Promise<ManifestSetCurrentVersionResult[]> {
  return invoke<ManifestSetCurrentVersionResult[]>('sourceUpdateRemoteVersions', { items });
}

async function stageRemoteBundle(
  bundleName: string,
  data: ManifestStageData
): Promise<ManifestStageResult> {
  return invoke<ManifestStageResult>('sourceStageRemoteBundle', { bundleName, data });
}

async function stageRemoteBundles(
  items: Record<string, ManifestStageData>
): Promise<ManifestStageResult[]> {
  return invoke<ManifestStageResult[]>('sourceStageRemoteBundles', { items });
}

async function removeRemoteBundle(
  bundleName: string,
  version: string,
  force?: boolean
): Promise<ManifestRemoveResult> {
  return invoke<ManifestRemoveResult>('sourceRemoveRemoteBundle', { bundleName, version, force });
}

async function removeRemoteBundles(
  items: Record<string, ManifestRemoveData>
): Promise<ManifestRemoveResult[]> {
  return invoke<ManifestRemoveResult[]>('sourceRemoveRemoteBundles', { items });
}

async function pruneRemoteBundle(bundleName: string): Promise<ManifestPruneResult> {
  return invoke<ManifestPruneResult>('sourcePruneRemoteBundle', { bundleName });
}

async function pruneRemoteBundles(bundleNames: string[]): Promise<ManifestPruneResult[]> {
  return invoke<ManifestPruneResult[]>('sourcePruneRemoteBundles', { bundleNames });
}

async function resolveFilepath(bundleName: string): Promise<string> {
  return invoke<string>('sourceResolveFilepath', { bundleName });
}

async function getBuiltinBundleFilepath(bundleName: string, version: string): Promise<string> {
  return invoke<string>('sourceGetBuiltinBundleFilepath', { bundleName, version });
}

async function getRemoteBundleFilepath(bundleName: string, version: string): Promise<string> {
  return invoke<string>('sourceGetRemoteBundleFilepath', { bundleName, version });
}

async function unload(bundleName: string): Promise<boolean> {
  return invoke<boolean>('sourceUnload', { bundleName });
}

export interface SourceApi {
  listBundles: typeof listBundles;
  listBuiltinBundles: typeof listBuiltinBundles;
  listRemoteBundles: typeof listRemoteBundles;
  getVersion: typeof getVersion;
  getRemoteStagedVersion: typeof getRemoteStagedVersion;
  getRemotePreviousVersion: typeof getRemotePreviousVersion;
  getBuiltinVersionData: typeof getBuiltinVersionData;
  getRemoteVersionData: typeof getRemoteVersionData;
  updateRemoteVersion: typeof updateRemoteVersion;
  updateRemoteVersions: typeof updateRemoteVersions;
  stageRemoteBundle: typeof stageRemoteBundle;
  stageRemoteBundles: typeof stageRemoteBundles;
  removeRemoteBundle: typeof removeRemoteBundle;
  removeRemoteBundles: typeof removeRemoteBundles;
  pruneRemoteBundle: typeof pruneRemoteBundle;
  pruneRemoteBundles: typeof pruneRemoteBundles;
  resolveFilepath: typeof resolveFilepath;
  getBuiltinBundleFilepath: typeof getBuiltinBundleFilepath;
  getRemoteBundleFilepath: typeof getRemoteBundleFilepath;
  unload: typeof unload;
}

export const source: SourceApi = {
  listBundles,
  listBuiltinBundles,
  listRemoteBundles,
  getVersion,
  getRemoteStagedVersion,
  getRemotePreviousVersion,
  getBuiltinVersionData,
  getRemoteVersionData,
  updateRemoteVersion,
  updateRemoteVersions,
  stageRemoteBundle,
  stageRemoteBundles,
  removeRemoteBundle,
  removeRemoteBundles,
  pruneRemoteBundle,
  pruneRemoteBundles,
  resolveFilepath,
  getBuiltinBundleFilepath,
  getRemoteBundleFilepath,
  unload,
};

File: packages/tauri/src/command/source.rs
use crate::WebviewBundleExtra;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tauri::{AppHandle, Runtime, command};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestVersionData {
  pub integrity: Option<String>,
  pub metadata: Option<HashMap<String, String>>,
}

impl From<wvb::source::ManifestVersionData> for ManifestVersionData {
  fn from(value: wvb::source::ManifestVersionData) -> Self {
    Self {
      integrity: value.integrity.clone(),
      metadata: value.metadata.clone(),
    }
  }
}

impl From<ManifestVersionData> for wvb::source::ManifestVersionData {
  fn from(value: ManifestVersionData) -> Self {
    Self {
      integrity: value.integrity,
      metadata: value.metadata,
    }
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ManifestBundleItemStatus {
  Current,
  Previous,
  Staged,
  Orphan,
}

impl From<wvb::source::ManifestBundleItemStatus> for ManifestBundleItemStatus {
  fn from(value: wvb::source::ManifestBundleItemStatus) -> Self {
    match value {
      wvb::source::ManifestBundleItemStatus::Current => Self::Current,
      wvb::source::ManifestBundleItemStatus::Previous => Self::Previous,
      wvb::source::ManifestBundleItemStatus::Staged => Self::Staged,
      wvb::source::ManifestBundleItemStatus::Orphan => Self::Orphan,
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestBundleItem {
  pub name: String,
  pub version: String,
  pub status: ManifestBundleItemStatus,
  pub data: ManifestVersionData,
}

impl From<wvb::source::ManifestBundleItem> for ManifestBundleItem {
  fn from(value: wvb::source::ManifestBundleItem) -> Self {
    Self {
      name: value.name,
      version: value.version,
      status: value.status.into(),
      data: value.data.into(),
    }
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SourceKind {
  Builtin,
  Remote,
}

impl From<wvb::source::SourceKind> for SourceKind {
  fn from(value: wvb::source::SourceKind) -> Self {
    match value {
      wvb::source::SourceKind::Builtin => SourceKind::Builtin,
      wvb::source::SourceKind::Remote => SourceKind::Remote,
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestListItem {
  pub source: SourceKind,
  #[serde(flatten)]
  pub item: ManifestBundleItem,
}

impl From<wvb::source::SourceListItem> for ManifestListItem {
  fn from(value: wvb::source::SourceListItem) -> Self {
    Self {
      source: value.source.into(),
      item: value.item.into(),
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BundleSourceVersion {
  pub source: SourceKind,
  pub version: String,
}

impl From<wvb::source::BundleSourceVersion> for BundleSourceVersion {
  fn from(value: wvb::source::BundleSourceVersion) -> Self {
    Self {
      source: value.source.into(),
      version: value.version,
    }
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ManifestSetCurrentVersionResultKind {
  Settled,
  NotExists,
  VersionNotExists,
}

impl From<wvb::source::ManifestSetCurrentVersionResultKind>
  for ManifestSetCurrentVersionResultKind
{
  fn from(value: wvb::source::ManifestSetCurrentVersionResultKind) -> Self {
    match value {
      wvb::source::ManifestSetCurrentVersionResultKind::Settled => Self::Settled,
      wvb::source::ManifestSetCurrentVersionResultKind::NotExists => Self::NotExists,
      wvb::source::ManifestSetCurrentVersionResultKind::VersionNotExists => Self::VersionNotExists,
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestSetCurrentVersionResult {
  pub name: String,
  pub version: String,
  pub kind: ManifestSetCurrentVersionResultKind,
}

impl From<wvb::source::ManifestSetCurrentVersionResult> for ManifestSetCurrentVersionResult {
  fn from(value: wvb::source::ManifestSetCurrentVersionResult) -> Self {
    Self {
      name: value.name,
      version: value.version,
      kind: value.kind.into(),
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestStageData {
  pub version: String,
  pub data: Option<ManifestVersionData>,
}

impl From<ManifestStageData> for wvb::source::ManifestStageData {
  fn from(value: ManifestStageData) -> Self {
    Self {
      version: value.version,
      data: value.data.map(Into::into),
    }
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ManifestStageResultKind {
  Staged,
  InUse,
}

impl From<wvb::source::ManifestStageResultKind> for ManifestStageResultKind {
  fn from(value: wvb::source::ManifestStageResultKind) -> Self {
    match value {
      wvb::source::ManifestStageResultKind::Staged => Self::Staged,
      wvb::source::ManifestStageResultKind::InUse => Self::InUse,
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestStageResult {
  pub name: String,
  pub version: String,
  pub kind: ManifestStageResultKind,
}

impl From<wvb::source::ManifestStageResult> for ManifestStageResult {
  fn from(value: wvb::source::ManifestStageResult) -> Self {
    Self {
      name: value.name,
      version: value.version,
      kind: value.kind.into(),
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestRemoveData {
  pub versions: Vec<String>,
  pub force: Option<bool>,
}

impl From<ManifestRemoveData> for wvb::source::ManifestRemoveData {
  fn from(value: ManifestRemoveData) -> Self {
    Self {
      versions: value.versions,
      force: value.force,
    }
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ManifestRemoveResultKind {
  Removed,
  NotExists,
  VersionNotExists,
  InUse,
}

impl From<wvb::source::ManifestRemoveResultKind> for ManifestRemoveResultKind {
  fn from(value: wvb::source::ManifestRemoveResultKind) -> Self {
    match value {
      wvb::source::ManifestRemoveResultKind::Removed => Self::Removed,
      wvb::source::ManifestRemoveResultKind::NotExists => Self::NotExists,
      wvb::source::ManifestRemoveResultKind::VersionNotExists => Self::VersionNotExists,
      wvb::source::ManifestRemoveResultKind::InUse => Self::InUse,
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestRemoveResult {
  pub name: String,
  pub version: String,
  pub kind: ManifestRemoveResultKind,
}

impl From<wvb::source::ManifestRemoveResult> for ManifestRemoveResult {
  fn from(value: wvb::source::ManifestRemoveResult) -> Self {
    Self {
      name: value.name,
      version: value.version,
      kind: value.kind.into(),
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestPruneResult {
  pub name: String,
  pub pruned_versions: Vec<String>,
}

impl From<wvb::source::ManifestPruneResult> for ManifestPruneResult {
  fn from(value: wvb::source::ManifestPruneResult) -> Self {
    Self {
      name: value.name,
      pruned_versions: value.pruned_versions,
    }
  }
}

#[command]
pub async fn source_list_bundles<R: Runtime>(
  app: AppHandle<R>,
) -> crate::Result<Vec<ManifestListItem>> {
  let wvb = app.wvb();
  let items = wvb
    .source()
    .list_bundles()
    .await?
    .into_iter()
    .map(ManifestListItem::from)
    .collect::<Vec<_>>();
  Ok(items)
}

#[command]
pub async fn source_list_builtin_bundles<R: Runtime>(
  app: AppHandle<R>,
) -> crate::Result<Vec<ManifestListItem>> {
  let wvb = app.wvb();
  let items = wvb
    .source()
    .list_builtin_bundles()
    .await?
    .into_iter()
    .map(ManifestListItem::from)
    .collect::<Vec<_>>();
  Ok(items)
}

#[command]
pub async fn source_list_remote_bundles<R: Runtime>(
  app: AppHandle<R>,
) -> crate::Result<Vec<ManifestListItem>> {
  let wvb = app.wvb();
  let items = wvb
    .source()
    .list_remote_bundles()
    .await?
    .into_iter()
    .map(ManifestListItem::from)
    .collect::<Vec<_>>();
  Ok(items)
}

#[command]
pub async fn source_get_version<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
) -> crate::Result<Option<BundleSourceVersion>> {
  let wvb = app.wvb();
  let version = wvb
    .source()
    .get_version(&bundle_name)
    .await?
    .map(BundleSourceVersion::from);
  Ok(version)
}

#[command]
pub async fn source_get_remote_staged_version<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
) -> crate::Result<Option<String>> {
  let wvb = app.wvb();
  let version = wvb.source().get_remote_staged_version(&bundle_name).await?;
  Ok(version)
}

#[command]
pub async fn source_get_remote_previous_version<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
) -> crate::Result<Option<String>> {
  let wvb = app.wvb();
  let version = wvb
    .source()
    .get_remote_previous_version(&bundle_name)
    .await?;
  Ok(version)
}

#[command]
pub async fn source_get_builtin_version_data<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
  version: String,
) -> crate::Result<Option<ManifestVersionData>> {
  let wvb = app.wvb();
  let data = wvb
    .source()
    .get_builtin_version_data(&bundle_name, &version)
    .await?
    .map(ManifestVersionData::from);
  Ok(data)
}

#[command]
pub async fn source_get_remote_version_data<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
  version: String,
) -> crate::Result<Option<ManifestVersionData>> {
  let wvb = app.wvb();
  let data = wvb
    .source()
    .get_remote_version_data(&bundle_name, &version)
    .await?
    .map(ManifestVersionData::from);
  Ok(data)
}

#[command]
pub async fn source_update_remote_version<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
  version: String,
) -> crate::Result<ManifestSetCurrentVersionResult> {
  let wvb = app.wvb();
  let result = wvb
    .source()
    .update_remote_version(&bundle_name, &version)
    .await?;
  Ok(result.into())
}

#[command]
pub async fn source_update_remote_versions<R: Runtime>(
  app: AppHandle<R>,
  items: HashMap<String, String>,
) -> crate::Result<Vec<ManifestSetCurrentVersionResult>> {
  let wvb = app.wvb();
  let results = wvb
    .source()
    .update_remote_versions(items)
    .await?
    .into_iter()
    .map(ManifestSetCurrentVersionResult::from)
    .collect::<Vec<_>>();
  Ok(results)
}

#[command]
pub async fn source_stage_remote_bundle<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
  data: ManifestStageData,
) -> crate::Result<ManifestStageResult> {
  let wvb = app.wvb();
  let result = wvb
    .source()
    .stage_remote_bundle(&bundle_name, data.into())
    .await?;
  Ok(result.into())
}

#[command]
pub async fn source_stage_remote_bundles<R: Runtime>(
  app: AppHandle<R>,
  items: HashMap<String, ManifestStageData>,
) -> crate::Result<Vec<ManifestStageResult>> {
  let wvb = app.wvb();
  let items = items
    .into_iter()
    .map(|(name, data)| (name, data.into()))
    .collect::<HashMap<String, wvb::source::ManifestStageData>>();
  let results = wvb
    .source()
    .stage_remote_bundles(items)
    .await?
    .into_iter()
    .map(ManifestStageResult::from)
    .collect::<Vec<_>>();
  Ok(results)
}

#[command]
pub async fn source_remove_remote_bundle<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
  version: String,
  force: Option<bool>,
) -> crate::Result<ManifestRemoveResult> {
  let wvb = app.wvb();
  let result = wvb
    .source()
    .remove_remote_bundle(&bundle_name, &version, force)
    .await?;
  Ok(result.into())
}

#[command]
pub async fn source_remove_remote_bundles<R: Runtime>(
  app: AppHandle<R>,
  items: HashMap<String, ManifestRemoveData>,
) -> crate::Result<Vec<ManifestRemoveResult>> {
  let wvb = app.wvb();
  let items = items
    .into_iter()
    .map(|(name, data)| (name, data.into()))
    .collect::<HashMap<String, wvb::source::ManifestRemoveData>>();
  let results = wvb
    .source()
    .remove_remote_bundles(items)
    .await?
    .into_iter()
    .map(ManifestRemoveResult::from)
    .collect::<Vec<_>>();
  Ok(results)
}

#[command]
pub async fn source_prune_remote_bundle<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
) -> crate::Result<ManifestPruneResult> {
  let wvb = app.wvb();
  let result = wvb.source().prune_remote_bundle(&bundle_name).await?;
  Ok(result.into())
}

#[command]
pub async fn source_prune_remote_bundles<R: Runtime>(
  app: AppHandle<R>,
  bundle_names: Vec<String>,
) -> crate::Result<Vec<ManifestPruneResult>> {
  let wvb = app.wvb();
  let results = wvb
    .source()
    .prune_remote_bundles(&bundle_names)
    .await?
    .into_iter()
    .map(ManifestPruneResult::from)
    .collect::<Vec<_>>();
  Ok(results)
}

#[command]
pub async fn source_resolve_filepath<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
) -> crate::Result<String> {
  let wvb = app.wvb();
  let filepath = wvb.source().resolve_filepath(&bundle_name).await?;
  Ok(filepath.to_string_lossy().to_string())
}

#[command]
pub async fn source_get_builtin_bundle_filepath<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
  version: String,
) -> crate::Result<String> {
  let wvb = app.wvb();
  let filepath = wvb
    .source()
    .get_builtin_bundle_filepath(&bundle_name, &version)?;
  Ok(filepath.to_string_lossy().to_string())
}

#[command]
pub async fn source_get_remote_bundle_filepath<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
  version: String,
) -> crate::Result<String> {
  let wvb = app.wvb();
  let filepath = wvb
    .source()
    .get_remote_bundle_filepath(&bundle_name, &version)?;
  Ok(filepath.to_string_lossy().to_string())
}

#[command]
pub async fn source_unload<R: Runtime>(
  app: AppHandle<R>,
  bundle_name: String,
) -> crate::Result<bool> {
  let wvb = app.wvb();
  Ok(wvb.source().unload(&bundle_name))
}
